Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/GlobalLocalAnomalyDetector.java

```diff
@@ -211,19 +214,35 @@ public GenericAnomalyDescriptor<P> process(P object, float weight, BiFunction<P,
                         globalDistance, null);
             }
         }
-        Weighted<P> result = score(object, localDistance, considerOcclusion);
-
+        List<Weighted<P>> result = score(object, localDistance, considerOcclusion);
         double threshold = thresholder.threshold();
         double grade = 0;
-        if (result.weight > 0) {
-            grade = thresholder.getAnomalyGrade(result.weight, false);
-        }
+        float score = 0;
+        if (result != null) {
+            score = result.stream().map(a -> a.weight).reduce(FLOAT_MAX, Float::min);
+            if (score < FLOAT_MAX) {
+                // an exponential attribution
+                double sum = result.stream()
+                        .map(a -> (double) ((a.weight == FLOAT_MAX) ? 0 : exp(-a.weight * a.weight)))
+                        .reduce(0.0, Double::sum);
+                for (Weighted<P> item : result) {
+                    item.weight = (item.weight == FLOAT_MAX) ? 0.0f
+                            : (float) min(1.0f, (float) exp(-item.weight * item.weight) / sum);
+                }
+            } else {
+                // uniform attribution
+                for (Weighted<P> item : result) {
+                    item.weight = (float) 1.0 / (result.size());
+                }
+            }
+            grade = thresholder.getAnomalyGrade(score, false);
 
+        }
         // note average score would be 1
-        thresholder.update(result.weight, min(result.weight, thresholder.getZFactor()));
+        thresholder.update(score, min(score, thresholder.getZFactor()));
         sample(object, weight);
 
-        return new GenericAnomalyDescriptor<>(result.index, result.weight, threshold, grade);
+        return new GenericAnomalyDescriptor<>(result, score, threshold, grade);
```

Comment:
AD may need to support the ability to find anomalies based on a feature which has unique values. For example, if a feature called error message has multiple messages such as “ Failed to process 123”, “Failed to process 234”, we should be able to detect anomalies for “Failed to process XXX” and not on the unique values. Can we use each representative's weight to find which cluster(s) the input object belongs? We can find largest weights with similar magnitude (e.g., no more than 10% error margin). The corresponding representative(s) are what we need to category where the log message belongs. Say we have 5 weights: 40% 38% 10% 10% 2%. 10% error margin is 40% * 0.1 = 4%. So we take the first two weights (since 38% is within the error margin of 40%), increment number of occurrences of the corresponding representatives by 1. In each interval, we will send the final count of each representatives to HCAD. Of course, we should have a minimum weights (e.g., 1.5/result.size()) as a threshold of even starting the vetting process.

Reply:
hello

Reply:
The representatives will allow some of this computation to proceed - but the feature generalization of {123,234} -> ZZZ has to be domain specific and should be done in the code that is calling this subroutine.